core/lib/zksync_core/src/consensus/storage/mod.rs

```diff
@@ -0,0 +1,394 @@
+//! Storage implementation based on DAL.
+use crate::consensus;
+use crate::consensus::wrap_error::WrapError;
+use anyhow::Context as _;
+use std::ops;
+use zksync_concurrency::{ctx, sync, time};
+use zksync_consensus_bft::PayloadSource;
+use zksync_consensus_roles::validator;
+use zksync_consensus_storage::{
+    BlockStore, ReplicaState, ReplicaStateStore, StorageError, StorageResult, WriteBlockStore,
+};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_types::{api::en::SyncBlock, block::ConsensusBlockFields, Address, MiniblockNumber};
+
+impl WrapError for StorageError {
+    fn with_wrap<C: std::fmt::Display + Send + Sync + 'static, F: FnOnce() -> C>(
+        self,
+        f: F,
+    ) -> Self {
+        match self {
+            Self::Database(err) => Self::Database(err.with_wrap(f)),
+            err => err,
+        }
+    }
+}
+
+pub(crate) fn sync_block_to_consensus_block(
+    mut block: SyncBlock,
+) -> anyhow::Result<validator::FinalBlock> {
+    let number = validator::BlockNumber(block.number.0.into());
+    let consensus = block.consensus.take().context("Missing consensus fields")?;
+    let payload: consensus::Payload = block.try_into()?;
+    let payload = payload.encode();
+    Ok(validator::FinalBlock {
+        header: validator::BlockHeader {
+            parent: consensus.parent,
+            number,
+            payload: payload.hash(),
+        },
+        payload,
+        justification: consensus.justification,
+    })
+}
+
+async fn storage<'a>(
+    ctx: &ctx::Ctx,
+    pool: &'a ConnectionPool,
+) -> StorageResult<StorageProcessor<'a>> {
+    ctx.wait(pool.access_storage_tagged("sync_layer"))
+        .await?
+        .map_err(StorageError::Database)
+}
+
+async fn start_transaction<'a, 'b, 'c: 'b>(
+    ctx: &ctx::Ctx,
+    storage: &'c mut StorageProcessor<'a>,
+) -> Result<StorageProcessor<'b>, StorageError> {
+    ctx.wait(storage.start_transaction())
+        .await?
+        .map_err(|err| StorageError::Database(err.into()))
+}
+
+async fn commit<'a>(ctx: &ctx::Ctx, txn: StorageProcessor<'a>) -> StorageResult<()> {
```

Comment:
Nit: `'a` can be elided here and in `fetch_block` (i.e., you can write `txn: StorageProcessor<'_>`).